File: Lab5/Student.cs
using System;

namespace struct_lab_student
{
    struct Student
    {
        public string surName;
        public string firstName;
        public string patronymic;
        public char sex;
        public string dateOfBirth;
        public char mathematicsMark;
        public char physicsMark;
        public char informaticsMark;
        public int scholarship;
        public double[] allMarks;

        public Student(string lineWithAllData)
        {
            // TODO   you SHOULD IMPLEMENT constructor with exactly this signature
            // lineWithAllData is string contating all data about one student, as described in statement
            string[] data = lineWithAllData.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);

            surName    = data[0];
            firstName  = data[1];
            patronymic = data[2];

            sex = Convert.ToChar(data[3]);

            dateOfBirth = data[4];

            char[] marks = { Convert.ToChar(data[5]), Convert.ToChar(data[6]), Convert.ToChar(data[7]) };
            double[] dMarks = new double[3];

            for(int i = 0; i < marks.Length; i++)
            {
                marks[i] = !char.IsNumber(marks[i]) 
                           || marks[i].Equals('-') ? '2' : 
                              marks[i] > 53 ? '5'        : marks[i];

                dMarks[i] = char.GetNumericValue(marks[i]);
            }

            allMarks = dMarks;

            mathematicsMark = marks[0];
            physicsMark     = marks[1];
            informaticsMark = marks[2];

            scholarship = int.TryParse(data[8], out scholarship) && scholarship > 122  ? scholarship :
                                                                    scholarship > 2346 ? 2345        : 0;
        }
    }
}

File: src/Lab5/Student.java
package Lab5;

import static java.lang.Integer.parseInt;

public class Student {
    String surName;
    String firstName;
    String patronymic;
    char sex;
    String dateOfBirth;
    char mathematicsMark;
    char physicsMark;
    char informaticsMark;
    int scholarship;

    public Student(String lineWithAllData) {
        String[] data = lineWithAllData.split(" +");

        surName = data[0];
        firstName = data[1];
        patronymic = data[2];

        sex = data[3].charAt(0);

        dateOfBirth = data[4];

        char[] marks = {data[5].charAt(0), data[6].charAt(0), data[7].charAt(0)};

        for (int i = 0; i < marks.length; i++) {
            if (!Character.isDigit(marks[i]) || marks[i] == '-') marks[i] = '2';
        }

        mathematicsMark = marks[0];
        physicsMark     = marks[1];
        informaticsMark = marks[2];

        try { scholarship = parseInt(data[8]);
        } catch (NumberFormatException e) { scholarship = 0; }
    }
}
